### DISCUSSION TITLE
uniform struct

### DISCUSSION THREAD

**Discussion by hejian01** [vsg-dev/VulkanSceneGraph#848] [Q&A] [answered]
I have attempted two approaches to pass a uniform struct to a shader, and the first one failed while the second one succeeded. My goal is to dynamically change the values in this uniform struct during the rendering loop. I believe the first approach is more in line with the VSG (VulkanSceneGraph) specification, but I'm unsure how to modify it. I would appreciate any suggestions or advice from experts.
1. First method
`struct PlaneInfo
{
vsg::ref_ptr<vsg::ivec4Value> m_indeices;
vsg::ref_ptr<vsg::vec4Value> m_color;
};

PlaneInfo planeInfo;
planeInfo.m_indeices = vsg::ivec4Value::create(0, 0, 0, 0);
planeInfo.m_indeices->properties.dataVariance = vsg::DYNAMIC_DATA;
planeInfo.m_color = vsg::vec4Value::create(0.0f, 0.0f, 1.0f, 1.0f);
planeInfo.m_color->properties.dataVariance = vsg::DYNAMIC_DATA;

auto descriptorBuffer = vsg::DescriptorBuffer::create(DataList{ planeInfo.m_indeices, planeInfo.m_color}, 0);

auto descriptorSet = vsg::DescriptorSet::create(descriptorSetLayout, vsg::Descriptors{ descriptorBuffer });
auto uniform_bindDescriptorSet = vsg::BindDescriptorSet::create(VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, descriptorSet);`

2. Second method:
`auto device = g_window->getOrCreateDevice();

auto bufferSize = sizeof(PlaneInfo);
auto uniformBuffer = vsg::createBufferAndMemory(device, bufferSize,
VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT,
VK_SHARING_MODE_EXCLUSIVE,
VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);

auto uniformBufferInfo = vsg::BufferInfo::create();
uniformBufferInfo->buffer = uniformBuffer;
uniformBufferInfo->offset = 0;
uniformBufferInfo->range = bufferSize;

PlaneInfo planeInfo;
planeInfo.m_indeices = vsg::ivec4Value::create(0, 0, 0, 0);
planeInfo.m_indeices->properties.dataVariance = vsg::DYNAMIC_DATA;
planeInfo.m_color = vsg::vec4Value::create(0.0f, 0.0f, 1.0f, 1.0f);
planeInfo.m_color->properties.dataVariance = vsg::DYNAMIC_DATA;

void* data;
uint32_t size = sizeof(PlaneInfo);
vkMapMemory(device->vk(), uniformBuffer->getDeviceMemory(device->deviceID)->vk(), 0, size, 0, &data);
memcpy(data, &planeInfo, sizeof(PlaneInfo));
vkUnmapMemory(device->vk(), uniformBuffer->getDeviceMemory(device->deviceID)->vk());

auto descriptorBuffer = vsg::DescriptorBuffer::create(BufferInfoList{ uniformBufferInfo }, 0, 0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER);

auto descriptorSet = vsg::DescriptorSet::create(descriptorSetLayout, vsg::Descriptors{ descriptorBuffer });
auto uniform_bindDescriptorSet = vsg::BindDescriptorSet::create(VK_PIPELINE_BIND_POINT_GRAPHICS, pipelineLayout, 0, descriptorSet);`

**Comment by vsg-dev** [comment, marked as answer]
I don't know why you have the PlaneInfo defined in case 1, it should be irrelevant to the operation.  However, in general just assigning the Data to DescrptoriBuffer as you have done should be OK, and they only bit missing will be calling data->dirty().  You don't say whether you've done this or not, or in what way it doesn't work.

The second doesn't included PlateInfo, is it the same definition as case 1? If so I can't see how passing pointers to objects in main memory will ever work when passing data to the GPU.

Could it be that case 1 actually worked and case 2 failed and just typed things up wrong?

My recommendation is to have a look at the [vsgclip](https://github.com/vsg-dev/vsgExamples/blob/master/examples/state/vsgclip/vsgclip.cpp#L249) example, this illustrates passing dynamically updated uniform data.

It is possible to allocate a buffer and then map it and copy to it, but you'll need to think about buffering of the copy region to avoid updating data while it's still being rendered by a previous frame, and the map/unmap is expensive.  The vsg::TransferData class does this buffering for you and avoids the map/unmap. as well as being automatically set up for you, so it's more robust, faster and simpler than manually managing and mapping the buffer.

When copying packed data you can use your own struct {} and then attach this to the DescriptorBuffer as data via vsg::Value<MyStruct> template.  This is how material structs are passed to the GPU.  See [include/vsg/state/material.h](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/state/material.h#L22) to see how this is done.

**Reply by hejian01** [reply]
Thank you very much. I referred to "matreial.h" to modify my uniform struct, and the final result is correct.

**Comment by WdsPjm** [comment]
I loaded the vsgt model in the vsgclip, passing uniform to the gpu was fine. But loading gltf model can not pass uniform to gpu, May I ask have encountered such a problem

**Reply by robertosfield** [reply]
vsgclip example works in a specific way that isn't generally compatible with range of models, it won't be that uniforms aren't being passed it'll be that the what the example is trying to do isn't compatible.

Rather than focus on the limitations of vsgclip it's better to tell us what you are attempting to do and what issues you are seeing.

**Comment by WdsPjm** [comment]
I want to add some functionality to vsg's default shader, or write a shader to replace vsg's default shader

**Reply by robertosfield** [reply]
I've started writing a new vsgExample that illustrates how to write a custom vsg::ShaderSet, this will show extend an existing ShaderSet to your own shaders and add extra uniforms.  This example isn't ready to merge with vsgExamples master but I'm hoping to complete it today.  

As part of this work I'm also generalizing ShaderSet/GraphicsPipelineConfigurator to make it more flexible.

**Comment by WdsPjm** [comment]
Great. Thank you so much. I'm looking forward to it.

**Reply by robertosfield** [reply]
For different topics it's best to start a dedicated thread.

**Reply by robertosfield** [reply]
The new example is still in development but is already far enough along to load a GLTF file and use custom fragment shader and uniform that isn't part of the standard PBR ShaderSet.  The branch is:

https://github.com/vsg-dev/vsgExamples/tree/CustomShaderSet

The example itself is:

https://github.com/vsg-dev/vsgExamples/tree/CustomShaderSet/examples/utils/vsgcustomshaderset

There are also new vsgExamples/data/shaders/custom_pbr.vert and custom_pbr.frag shaders.

**Comment by WdsPjm** [comment]
Robert Sfield, you are amazing
